README.md
# Work in progress. Please, don't try to use it.

This branch will become v2 in the future, but now it is possible rebases,
force-pushes etc.

<a href="http://tarantool.org">
	<img src="https://avatars2.githubusercontent.com/u/2344919?v=2&s=250" align="right">
</a>

[![Go Reference][godoc-badge]][godoc-url]
[![Actions Status][actions-badge]][actions-url]
[![Code Coverage][coverage-badge]][coverage-url]
[![Telegram][telegram-badge]][telegram-url]
[![GitHub Discussions][discussions-badge]][discussions-url]
[![Stack Overflow][stackoverflow-badge]][stackoverflow-url]

# Client in Go for Tarantool

The package `go-tarantool` contains everything you need to connect to
[Tarantool 1.6+][tarantool-site].

The advantage of integrating Go with Tarantool, which is an application server
plus a DBMS, is that Go programmers can handle databases and perform on-the-fly
recompilations of embedded Lua routines, just as in C, with responses that are
faster than other packages according to public benchmarks.

## Table of contents

* [Installation](#installation)
  * [Build tags](#build-tags)
* [Documentation](#documentation)
  * [API reference](#api-reference)
  * [Walking\-through example](#walking-through-example)
  * [msgpack.v5 migration](#msgpackv5-migration)
* [Contributing](#contributing)
* [Alternative connectors](#alternative-connectors)

## Installation

We assume that you have Tarantool version 1.6+ and a modern Linux or BSD
operating system.

You need a current version of `go`, version 1.13 or later (use `go version` to
check the version number). Do not use `gccgo-go`.

**Note:** If your `go` version is older than 1.13 or if `go` is not installed,
download and run the latest tarball from [golang.org][golang-dl].

The package `go-tarantool` is located in [tarantool/go-tarantool][go-tarantool]
repository. To download and install, say:

```
$ go get github.com/tarantool/go-tarantool
```

This should put the source and binary files in subdirectories of
`/usr/local/go`, so that you can access them by adding
`github.com/tarantool/go-tarantool` to the `import {...}` section at the start
of any Go program.

### Build tags

We define multiple [build tags](https://pkg.go.dev/go/build#hdr-Build_Constraints).

This allows us to introduce new features without losing backward compatibility.

1. To disable SSL support and linking with OpenSSL, you can use the tag:
   ```
   go_tarantool_ssl_disable
   ```
2. To change the default `Call` behavior from `Call16` to `Call17`, you can use
   the build tag:
   ```
   go_tarantool_call_17
   ```
   **Note:** In future releases, `Call17` may be used as default `Call` behavior.
3. To replace usage of `msgpack.v2` with `msgpack.v5`, you can use the build
   tag:
   ```
   go_tarantool_msgpack_v5
   ```
   **Note:** In future releases, `msgpack.v5` may be used by default. We recommend
   to read [msgpack.v5 migration notes](#msgpackv5-migration) and try to
   use msgpack.v5 before the changes.
4. To run fuzz tests with decimals, you can use the build tag:
   ```
   go_tarantool_decimal_fuzzing
   ```
   **Note:** It crashes old Tarantool versions and requires Go 1.18+.

## Documentation

Read the [Tarantool documentation][tarantool-doc-data-model-url]
to find descriptions of terms such as "connect", "space", "index", and the
requests to create and manipulate database objects or Lua functions.

In general, connector methods can be divided into two main parts:

* `Connect()` function and functions related to connecting, and
* Data manipulation functions and Lua invocations such as `Insert()` or `Call()`.

The supported requests have parameters and results equivalent to requests in
the [Tarantool CRUD operations][tarantool-doc-box-space-url].
There are also Typed and Async versions of each data-manipulation function.

### API Reference

Learn API documentation and examples at [pkg.go.dev][godoc-url].

### Walking-through example

We can now have a closer look at the example and make some observations
about what it does.

```go
package tarantool

import (
	"fmt"
	"github.com/tarantool/go-tarantool"
)

func main() {
	opts := tarantool.Opts{User: "guest"}
	conn, err := tarantool.Connect("127.0.0.1:3301", opts)
	if err != nil {
		fmt.Println("Connection refused:", err)
	}
	resp, err := conn.Insert(999, []interface{}{99999, "BB"})
	if err != nil {
		fmt.Println("Error", err)
		fmt.Println("Code", resp.Code)
	}
}
```

**Observation 1:** The line "`github.com/tarantool/go-tarantool`" in the
`import(...)` section brings in all Tarantool-related functions and structures.

**Observation 2:** The line starting with "`Opts :=`" sets up the options for
`Connect()`. In this example, the structure contains only a single value, the
username. The structure may also contain other settings, see more in
[documentation][godoc-opts-url] for the "`Opts`" structure.

**Observation 3:** The line containing "`tarantool.Connect`" is essential for
starting a session. There are two parameters:

* a string with `host:port` format, and
* the option structure that was set up earlier.

**Observation 4:** The `err` structure will be `nil` if there is no error,
otherwise it will have a description which can be retrieved with `err.Error()`.

**Observation 5:** The `Insert` request, like almost all requests, is preceded by
"`conn.`" which is the name of the object that was returned by `Connect()`.
There are two parameters:

* a space number (it could just as easily have been a space name), and
* a tuple.

### msgpack.v5 migration

Most function names and argument types in `msgpack.v5` and `msgpack.v2`
have not changed (in our code, we noticed changes in `EncodeInt`, `EncodeUint`
and `RegisterExt`). But there are a lot of changes in a logic of encoding and
decoding. On the plus side the migration seems easy, but on the minus side you
need to be very careful.

First of all, `EncodeInt8`, `EncodeInt16`, `EncodeInt32`, `EncodeInt64`
and `EncodeUint*` analogues at `msgpack.v5` encode numbers as is without loss of
type. In `msgpack.v2` the type of a number is reduced to a value.

Secondly, a base decoding function does not convert numbers to `int64` or
`uint64`. It converts numbers to an exact type defined by MessagePack. The
change makes manual type conversions much more difficult and can lead to
runtime errors with an old code. We do not recommend to use type conversions
and give preference to `*Typed` functions (besides, it's faster).

There are also changes in the logic that can lead to errors in the old code,
[as example](https://github.com/vmihailenco/msgpack/issues/327). Although in
`msgpack.v5` some functions for the logic tuning were added (see
[UseLooseInterfaceDecoding](https://pkg.go.dev/github.com/vmihailenco/msgpack/v5#Decoder.UseLooseInterfaceDecoding), [UseCompactInts](https://pkg.go.dev/github.com/vmihailenco/msgpack/v5#Encoder.UseCompactInts) etc), it is still impossible
to achieve full compliance of behavior between `msgpack.v5` and `msgpack.v2`. So
we don't go this way. We use standard settings if it possible.

## Contributing

See [the contributing guide](CONTRIBUTING.md) for detailed instructions on how
to get started with our project.

## Alternative connectors

There are two other connectors available from the open source community:

* [viciious/go-tarantool](https://github.com/viciious/go-tarantool),
* [FZambia/tarantool](https://github.com/FZambia/tarantool).

See feature comparison in the [documentation][tarantool-doc-connectors-comparison].

[tarantool-site]: https://tarantool.io/
[godoc-badge]: https://pkg.go.dev/badge/github.com/tarantool/go-tarantool.svg
[godoc-url]: https://pkg.go.dev/github.com/tarantool/go-tarantool
[actions-badge]: https://github.com/tarantool/go-tarantool/actions/workflows/testing.yml/badge.svg
[actions-url]: https://github.com/tarantool/go-tarantool/actions/workflows/testing.yml
[coverage-badge]: https://coveralls.io/repos/github/tarantool/go-tarantool/badge.svg?branch=master
[coverage-url]: https://coveralls.io/github/tarantool/go-tarantool?branch=master
[telegram-badge]: https://img.shields.io/badge/Telegram-join%20chat-blue.svg
[telegram-url]: http://telegram.me/tarantool
[discussions-badge]: https://img.shields.io/github/discussions/tarantool/tarantool
[discussions-url]: https://github.com/tarantool/tarantool/discussions
[stackoverflow-badge]: https://img.shields.io/badge/stackoverflow-tarantool-orange.svg
[stackoverflow-url]: https://stackoverflow.com/questions/tagged/tarantool
[golang-dl]: https://go.dev/dl/
[go-tarantool]: https://github.com/tarantool/go-tarantool
[tarantool-doc-data-model-url]: https://www.tarantool.io/en/doc/latest/book/box/data_model/
[tarantool-doc-box-space-url]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/
[godoc-opts-url]: https://pkg.go.dev/github.com/tarantool/go-tarantool#Opts
[tarantool-doc-connectors-comparison]: https://www.tarantool.io/en/doc/latest/book/connectors/#go-feature-comparison


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.

## [Unreleased]

### Added

- A Makefile target to test with race detector (#218)

### Changed

### Fixed

- Several non-critical data race issues (#218)

## [1.10.0] - 2022-12-31

The release improves compatibility with new Tarantool versions.

### Added

- Support iproto feature discovery (#120)
- Support errors extended information (#209)
- Support error type in MessagePack (#209)
- Support event subscription (#119)
- Support session settings (#215)
- Support pap-sha256 authorization method (Tarantool EE feature) (#243)
- Support graceful shutdown (#214)

### Fixed

- Decimal package uses a test variable DecimalPrecision instead of a
  package-level variable decimalPrecision (#233)
- Flaky test TestClientRequestObjectsWithContext (#244)
- Flaky test multi/TestDisconnectAll (#234)

## [1.9.0] - 2022-11-02

The release adds support for the latest version of the
[queue package](https://github.com/tarantool/queue) with master-replica
switching.

### Added

- Support the queue 1.2.1 (#177)
- ConnectionHandler interface for handling changes of connections in
  ConnectionPool (#178)
- Execute, ExecuteTyped and ExecuteAsync methods to ConnectionPool (#176)
- ConnectorAdapter type to use ConnectionPool as Connector interface (#176)
- An example how to use queue and connection_pool subpackages together (#176)

### Fixed

- Mode type description in the connection_pool subpackage (#208)
- Missed Role type constants in the connection_pool subpackage (#208)
- ConnectionPool does not close UnknownRole connections (#208)
- Segmentation faults in ConnectionPool requests after disconnect (#208)
- Addresses in ConnectionPool may be changed from an external code (#208)
- ConnectionPool recreates connections too often (#208)
- A connection is still opened after ConnectionPool.Close() (#208)
- Future.GetTyped() after Future.Get() does not decode response
  correctly (#213)
- Decimal package uses a test function GetNumberLength instead of a
  package-level function getNumberLength (#219)
- Datetime location after encode + decode is unequal (#217)
- Wrong interval arithmetic with timezones (#221)
- Invalid MsgPack if STREAM_ID > 127 (#224)
- queue.Take() returns an invalid task (#222)

## [1.8.0] - 2022-08-17

The minor release with time zones and interval support for datetime.

### Added

- Optional msgpack.v5 usage (#124)
- TZ support for datetime (#163)
- Interval support for datetime (#165)

### Changed

### Fixed

- Markdown of documentation for the decimal subpackage (#201)

## [1.7.0] - 2022-08-02

This release adds a number of features. The extending of the public API has
become possible with a new way of creating requests. New types of requests are
created via chain calls. Streams, context and prepared statements support are
based on this idea.

### Added

- SSL support (#155)
- IPROTO_PUSH messages support (#67)
- Public API with request object types (#126)
- Support decimal type in msgpack (#96)
- Support datetime type in msgpack (#118)
- Prepared SQL statements (#117)
- Context support for request objects (#48)
- Streams and interactive transactions support (#101)
- `Call16` method, support build tag `go_tarantool_call_17` to choose
  default behavior for `Call` method as Call17 (#125)

### Changed

- `IPROTO_*` constants that identify requests renamed from `<Name>Request` to
  `<Name>RequestCode` (#126)

### Removed

- NewErrorFuture function (#190)

### Fixed

- Add `ExecuteAsync` and `ExecuteTyped` to common connector interface (#62)

## [1.6.0] - 2022-06-01

This release adds a number of features. Also it significantly improves testing,
CI and documentation.

### Added

- Coveralls support (#149)
- Reusable testing workflow (integration testing with latest Tarantool) (#112)
- Simple CI based on GitHub actions (#114)
- Support UUID type in msgpack (#90)
- Go modules support (#91)
- queue-utube handling (#85)
- Master discovery (#113)
- SQL support (#62)

### Changed

- Handle everything with `go test` (#115)
- Use plain package instead of module for UUID submodule (#134)
- Reset buffer if its average use size smaller than quarter of capacity (#95)
- Update API documentation: comments and examples (#123)

### Fixed

- Fix queue tests (#107)
- Make test case consistent with comments (#105)

## [1.5] - 2019-12-29

First release.

### Fixed

- Fix infinite recursive call of `Upsert` method for `ConnectionMulti`
- Fix index out of range panic on `dial()` to short address
- Fix cast in `defaultLogger.Report` (#49)
- Fix race condition on extremely small request timeouts (#43)
- Fix notify for `Connected` transition
- Fix reconnection logic and add `Opts.SkipSchema` method
- Fix future sending
- Fix panic on disconnect + timeout
- Fix block on msgpack error
- Fix ratelimit
- Fix `timeouts` method for `Connection`
- Fix possible race condition on extremely small request timeouts
- Fix race condition on future channel creation
- Fix block on forever closed connection
- Fix race condition in `Connection`
- Fix extra map fields
- Fix response header parsing
- Fix reconnect logic in `Connection`

### Changed

- Make logger configurable
- Report user mismatch error immediately

multi/multi.go
// Package with methods to work with a Tarantool cluster.
//
// Main features:
//
// - Check the active connection with a configurable time interval and switch
// to the next connection in the pool if there is a connection failure.
//
// - Get the address list from the server and reconfigure it for use in
// MultiConnection.
//
// Since: 1.5
package multi

import (
	"errors"
	"fmt"
	"sync"
	"sync/atomic"
	"time"

	"github.com/tarantool/go-tarantool/v2"
)

const (
	connConnected = iota
	connClosed
)

var (
	ErrEmptyAddrs        = errors.New("addrs should not be empty")
	ErrWrongCheckTimeout = errors.New("wrong check timeout, must be greater than 0")
	ErrNoConnection      = errors.New("no active connections")
)

func indexOf(sstring string, data []string) int {
	for i, v := range data {
		if sstring == v {
			return i
		}
	}
	return -1
}

// ConnectionMulti is a handle with connections to a number of Tarantool instances.
//
// It is created and configured with Connect function, and could not be
// reconfigured later.
type ConnectionMulti struct {
	addrs    []string
	connOpts tarantool.Opts
	opts     OptsMulti

	mutex    sync.RWMutex
	notify   chan tarantool.ConnEvent
	state    uint32
	control  chan struct{}
	pool     map[string]*tarantool.Connection
	fallback *tarantool.Connection
}

var _ = tarantool.Connector(&ConnectionMulti{}) // Check compatibility with connector interface.

// OptsMulti is a way to configure Connection with multiconnect-specific options.
type OptsMulti struct {
	// CheckTimeout is a time interval to check for connection timeout and try to
	// switch connection.
	CheckTimeout time.Duration
	// Lua function name of the server called to retrieve the address list.
	NodesGetFunctionName string
	// Time interval to ask the server for an updated address list (works
	// if NodesGetFunctionName is set).
	ClusterDiscoveryTime time.Duration
}

// Connect creates and configures new ConnectionMulti with multiconnection options.
func ConnectWithOpts(addrs []string, connOpts tarantool.Opts, opts OptsMulti) (connMulti *ConnectionMulti, err error) {
	if len(addrs) == 0 {
		return nil, ErrEmptyAddrs
	}
	if opts.CheckTimeout <= 0 {
		return nil, ErrWrongCheckTimeout
	}
	if opts.ClusterDiscoveryTime <= 0 {
		opts.ClusterDiscoveryTime = 60 * time.Second
	}

	notify := make(chan tarantool.ConnEvent, 10*len(addrs)) // x10 to accept disconnected and closed event (with a margin).
	connOpts.Notify = notify
	connMulti = &ConnectionMulti{
		addrs:    addrs,
		connOpts: connOpts.Clone(),
		opts:     opts,
		notify:   notify,
		control:  make(chan struct{}),
		pool:     make(map[string]*tarantool.Connection),
	}
	somebodyAlive, _ := connMulti.warmUp()
	if !somebodyAlive {
		connMulti.Close()
		return nil, ErrNoConnection
	}
	go connMulti.checker()

	return connMulti, nil
}

// Connect creates and configures new ConnectionMulti.
func Connect(addrs []string, connOpts tarantool.Opts) (connMulti *ConnectionMulti, err error) {
	opts := OptsMulti{
		CheckTimeout: 1 * time.Second,
	}
	return ConnectWithOpts(addrs, connOpts, opts)
}

func (connMulti *ConnectionMulti) warmUp() (somebodyAlive bool, errs []error) {
	errs = make([]error, len(connMulti.addrs))

	for i, addr := range connMulti.addrs {
		conn, err := tarantool.Connect(addr, connMulti.connOpts)
		errs[i] = err
		if conn != nil && err == nil {
			if connMulti.fallback == nil {
				connMulti.fallback = conn
			}
			connMulti.pool[addr] = conn
			if conn.ConnectedNow() {
				somebodyAlive = true
			}
		}
	}
	return
}

func (connMulti *ConnectionMulti) getState() uint32 {
	return atomic.LoadUint32(&connMulti.state)
}

func (connMulti *ConnectionMulti) getConnectionFromPool(addr string) (*tarantool.Connection, bool) {
	connMulti.mutex.RLock()
	defer connMulti.mutex.RUnlock()
	conn, ok := connMulti.pool[addr]
	return conn, ok
}

func (connMulti *ConnectionMulti) setConnectionToPool(addr string, conn *tarantool.Connection) {
	connMulti.mutex.Lock()
	defer connMulti.mutex.Unlock()
	connMulti.pool[addr] = conn
}

func (connMulti *ConnectionMulti) deleteConnectionFromPool(addr string) {
	connMulti.mutex.Lock()
	defer connMulti.mutex.Unlock()
	delete(connMulti.pool, addr)
}

func (connMulti *ConnectionMulti) checker() {

	refreshTimer := time.NewTicker(connMulti.opts.ClusterDiscoveryTime)
	timer := time.NewTicker(connMulti.opts.CheckTimeout)
	defer refreshTimer.Stop()
	defer timer.Stop()

	for connMulti.getState() != connClosed {

		select {
		case <-connMulti.control:
			return
		case e := <-connMulti.notify:
			if connMulti.getState() == connClosed {
				return
			}
			if e.Conn.ClosedNow() {
				addr := e.Conn.Addr()
				if _, ok := connMulti.getConnectionFromPool(addr); !ok {
					continue
				}
				conn, _ := tarantool.Connect(addr, connMulti.connOpts)
				if conn != nil {
					connMulti.setConnectionToPool(addr, conn)
				} else {
					connMulti.deleteConnectionFromPool(addr)
				}
			}
		case <-refreshTimer.C:
			if connMulti.getState() == connClosed || connMulti.opts.NodesGetFunctionName == "" {
				continue
			}
			var resp [][]string
			err := connMulti.Call17Typed(connMulti.opts.NodesGetFunctionName, []interface{}{}, &resp)
			if err != nil {
				continue
			}
			if len(resp) > 0 && len(resp[0]) > 0 {
				addrs := resp[0]
				// Fill pool with new connections.
				for _, v := range addrs {
					if indexOf(v, connMulti.addrs) < 0 {
						conn, _ := tarantool.Connect(v, connMulti.connOpts)
						if conn != nil {
							connMulti.setConnectionToPool(v, conn)
						}
					}
				}
				// Clear pool from obsolete connections.
				for _, v := range connMulti.addrs {
					if indexOf(v, addrs) < 0 {
						con, ok := connMulti.getConnectionFromPool(v)
						if con != nil && ok {
							con.Close()
						}
						connMulti.deleteConnectionFromPool(v)
					}
				}
				connMulti.addrs = addrs
			}
		case <-timer.C:
			for _, addr := range connMulti.addrs {
				if connMulti.getState() == connClosed {
					return
				}
				if conn, ok := connMulti.getConnectionFromPool(addr); ok {
					if !conn.ClosedNow() {
						continue
					}
				}
				conn, _ := tarantool.Connect(addr, connMulti.connOpts)
				if conn != nil {
					connMulti.setConnectionToPool(addr, conn)
				}
			}
		}
	}
}

func (connMulti *ConnectionMulti) getCurrentConnection() *tarantool.Connection {
	connMulti.mutex.RLock()
	defer connMulti.mutex.RUnlock()

	for _, addr := range connMulti.addrs {
		conn := connMulti.pool[addr]
		if conn != nil {
			if conn.ConnectedNow() {
				return conn
			}
			connMulti.fallback = conn
		}
	}
	return connMulti.fallback
}

// ConnectedNow reports if connection is established at the moment.
func (connMulti *ConnectionMulti) ConnectedNow() bool {
	return connMulti.getState() == connConnected && connMulti.getCurrentConnection().ConnectedNow()
}

// Close closes Connection.
// After this method called, there is no way to reopen this Connection.
func (connMulti *ConnectionMulti) Close() (err error) {
	connMulti.mutex.Lock()
	defer connMulti.mutex.Unlock()

	close(connMulti.control)
	connMulti.state = connClosed

	for _, conn := range connMulti.pool {
		if err == nil {
			err = conn.Close()
		} else {
			conn.Close()
		}
	}
	if connMulti.fallback != nil {
		connMulti.fallback.Close()
	}

	return
}

// Ping sends empty request to Tarantool to check connection.
func (connMulti *ConnectionMulti) Ping() (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Ping()
}

// ConfiguredTimeout returns a timeout from connection config.
func (connMulti *ConnectionMulti) ConfiguredTimeout() time.Duration {
	return connMulti.getCurrentConnection().ConfiguredTimeout()
}

// Select performs select to box space.
func (connMulti *ConnectionMulti) Select(space, index interface{}, offset, limit, iterator uint32, key interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Select(space, index, offset, limit, iterator, key)
}

// Insert performs insertion to box space.
// Tarantool will reject Insert when tuple with same primary key exists.
func (connMulti *ConnectionMulti) Insert(space interface{}, tuple interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Insert(space, tuple)
}

// Replace performs "insert or replace" action to box space.
// If tuple with same primary key exists, it will be replaced.
func (connMulti *ConnectionMulti) Replace(space interface{}, tuple interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Replace(space, tuple)
}

// Delete performs deletion of a tuple by key.
// Result will contain array with deleted tuple.
func (connMulti *ConnectionMulti) Delete(space, index interface{}, key interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Delete(space, index, key)
}

// Update performs update of a tuple by key.
// Result will contain array with updated tuple.
func (connMulti *ConnectionMulti) Update(space, index interface{}, key, ops interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Update(space, index, key, ops)
}

// Upsert performs "update or insert" action of a tuple by key.
// Result will not contain any tuple.
func (connMulti *ConnectionMulti) Upsert(space interface{}, tuple, ops interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Upsert(space, tuple, ops)
}

// Call calls registered Tarantool function.
// It uses request code for Tarantool >= 1.7 if go-tarantool
// was build with go_tarantool_call_17 tag.
// Otherwise, uses request code for Tarantool 1.6.
func (connMulti *ConnectionMulti) Call(functionName string, args interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Call(functionName, args)
}

// Call16 calls registered Tarantool function.
// It uses request code for Tarantool 1.6, so result is converted to array of
// arrays.
// Deprecated since Tarantool 1.7.2.
func (connMulti *ConnectionMulti) Call16(functionName string, args interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Call16(functionName, args)
}

// Call17 calls registered Tarantool function.
// It uses request code for Tarantool >= 1.7, so result is not converted
// (though, keep in mind, result is always array).
func (connMulti *ConnectionMulti) Call17(functionName string, args interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Call17(functionName, args)
}

// Eval passes Lua expression for evaluation.
func (connMulti *ConnectionMulti) Eval(expr string, args interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Eval(expr, args)
}

// Execute passes sql expression to Tarantool for execution.
//
// Since 1.6.0
func (connMulti *ConnectionMulti) Execute(expr string, args interface{}) (resp *tarantool.Response, err error) {
	return connMulti.getCurrentConnection().Execute(expr, args)
}

// GetTyped performs select (with limit = 1 and offset = 0) to box space and
// fills typed result.
func (connMulti *ConnectionMulti) GetTyped(space, index interface{}, key interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().GetTyped(space, index, key, result)
}

// SelectTyped performs select to box space and fills typed result.
func (connMulti *ConnectionMulti) SelectTyped(space, index interface{}, offset, limit, iterator uint32, key interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().SelectTyped(space, index, offset, limit, iterator, key, result)
}

// InsertTyped performs insertion to box space.
// Tarantool will reject Insert when tuple with same primary key exists.
func (connMulti *ConnectionMulti) InsertTyped(space interface{}, tuple interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().InsertTyped(space, tuple, result)
}

// ReplaceTyped performs "insert or replace" action to box space.
// If tuple with same primary key exists, it will be replaced.
func (connMulti *ConnectionMulti) ReplaceTyped(space interface{}, tuple interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().ReplaceTyped(space, tuple, result)
}

// DeleteTyped performs deletion of a tuple by key and fills result with
// deleted tuple.
func (connMulti *ConnectionMulti) DeleteTyped(space, index interface{}, key interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().DeleteTyped(space, index, key, result)
}

// UpdateTyped performs update of a tuple by key and fills result with updated
// tuple.
func (connMulti *ConnectionMulti) UpdateTyped(space, index interface{}, key, ops interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().UpdateTyped(space, index, key, ops, result)
}

// CallTyped calls registered function.
// It uses request code for Tarantool >= 1.7 if go-tarantool
// was build with go_tarantool_call_17 tag.
// Otherwise, uses request code for Tarantool 1.6.
func (connMulti *ConnectionMulti) CallTyped(functionName string, args interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().CallTyped(functionName, args, result)
}

// Call16Typed calls registered function.
// It uses request code for Tarantool 1.6, so result is converted to array of
// arrays.
// Deprecated since Tarantool 1.7.2.
func (connMulti *ConnectionMulti) Call16Typed(functionName string, args interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().Call16Typed(functionName, args, result)
}

// Call17Typed calls registered function.
// It uses request code for Tarantool >= 1.7, so result is not converted (though,
// keep in mind, result is always array)
func (connMulti *ConnectionMulti) Call17Typed(functionName string, args interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().Call17Typed(functionName, args, result)
}

// EvalTyped passes Lua expression for evaluation.
func (connMulti *ConnectionMulti) EvalTyped(expr string, args interface{}, result interface{}) (err error) {
	return connMulti.getCurrentConnection().EvalTyped(expr, args, result)
}

// ExecuteTyped passes sql expression to Tarantool for execution.
func (connMulti *ConnectionMulti) ExecuteTyped(expr string, args interface{}, result interface{}) (tarantool.SQLInfo, []tarantool.ColumnMetaData, error) {
	return connMulti.getCurrentConnection().ExecuteTyped(expr, args, result)
}

// SelectAsync sends select request to Tarantool and returns Future.
func (connMulti *ConnectionMulti) SelectAsync(space, index interface{}, offset, limit, iterator uint32, key interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().SelectAsync(space, index, offset, limit, iterator, key)
}

// InsertAsync sends insert action to Tarantool and returns Future.
// Tarantool will reject Insert when tuple with same primary key exists.
func (connMulti *ConnectionMulti) InsertAsync(space interface{}, tuple interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().InsertAsync(space, tuple)
}

// ReplaceAsync sends "insert or replace" action to Tarantool and returns Future.
// If tuple with same primary key exists, it will be replaced.
func (connMulti *ConnectionMulti) ReplaceAsync(space interface{}, tuple interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().ReplaceAsync(space, tuple)
}

// DeleteAsync sends deletion action to Tarantool and returns Future.
// Future's result will contain array with deleted tuple.
func (connMulti *ConnectionMulti) DeleteAsync(space, index interface{}, key interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().DeleteAsync(space, index, key)
}

// Update sends deletion of a tuple by key and returns Future.
// Future's result will contain array with updated tuple.
func (connMulti *ConnectionMulti) UpdateAsync(space, index interface{}, key, ops interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().UpdateAsync(space, index, key, ops)
}

// UpsertAsync sends "update or insert" action to Tarantool and returns Future.
// Future's sesult will not contain any tuple.
func (connMulti *ConnectionMulti) UpsertAsync(space interface{}, tuple interface{}, ops interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().UpsertAsync(space, tuple, ops)
}

// CallAsync sends a call to registered Tarantool function and returns Future.
// It uses request code for Tarantool >= 1.7 if go-tarantool
// was build with go_tarantool_call_17 tag.
// Otherwise, uses request code for Tarantool 1.6.
func (connMulti *ConnectionMulti) CallAsync(functionName string, args interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().CallAsync(functionName, args)
}

// Call16Async sends a call to registered Tarantool function and returns Future.
// It uses request code for Tarantool 1.6, so future's result is always array
// of arrays.
// Deprecated since Tarantool 1.7.2.
func (connMulti *ConnectionMulti) Call16Async(functionName string, args interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().Call16Async(functionName, args)
}

// Call17Async sends a call to registered Tarantool function and returns Future.
// It uses request code for Tarantool >= 1.7, so future's result will not be converted
// (though, keep in mind, result is always array).
func (connMulti *ConnectionMulti) Call17Async(functionName string, args interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().Call17Async(functionName, args)
}

// EvalAsync passes Lua expression for evaluation.
func (connMulti *ConnectionMulti) EvalAsync(expr string, args interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().EvalAsync(expr, args)
}

// ExecuteAsync passes sql expression to Tarantool for execution.
func (connMulti *ConnectionMulti) ExecuteAsync(expr string, args interface{}) *tarantool.Future {
	return connMulti.getCurrentConnection().ExecuteAsync(expr, args)
}

// NewPrepared passes a sql statement to Tarantool for preparation synchronously.
func (connMulti *ConnectionMulti) NewPrepared(expr string) (*tarantool.Prepared, error) {
	return connMulti.getCurrentConnection().NewPrepared(expr)
}

// NewStream creates new Stream object for connection.
//
// Since v. 2.10.0, Tarantool supports streams and interactive transactions over them.
// To use interactive transactions, memtx_use_mvcc_engine box option should be set to true.
// Since 1.7.0
func (connMulti *ConnectionMulti) NewStream() (*tarantool.Stream, error) {
	return connMulti.getCurrentConnection().NewStream()
}

// NewWatcher does not supported by the ConnectionMulti. The ConnectionMulti is
// deprecated: use ConnectionPool instead.
//
// Since 1.10.0
func (connMulti *ConnectionMulti) NewWatcher(key string,
	callback tarantool.WatchCallback) (tarantool.Watcher, error) {
	return nil, errors.New("ConnectionMulti is deprecated " +
		"use ConnectionPool")
}

// Do sends the request and returns a future.
func (connMulti *ConnectionMulti) Do(req tarantool.Request) *tarantool.Future {
	if connectedReq, ok := req.(tarantool.ConnectedRequest); ok {
		_, belongs := connMulti.getConnectionFromPool(connectedReq.Conn().Addr())
		if !belongs {
			fut := tarantool.NewFuture()
			fut.SetError(fmt.Errorf("the passed connected request doesn't belong to the current connection or connection pool"))
			return fut
		}
		return connectedReq.Conn().Do(req)
	}
	return connMulti.getCurrentConnection().Do(req)
}
